### ISSUE TITLE
Tidy up the model subpackages

### ISSUE THREAD

**Issue by akihironitta** [PyTorchLightning/pytorch-lightning-bolts#427]
## 🚀 Refactoring
The structure in `pl_bolts.models` is not user-friendly due to:

- an inconsistent structure, for example:
   ```python
   from pl_bolts.models.vision import SemSegment
   from pl_bolts.models.vision import ImageGPT    # This doesn't work,
   from pl_bolts.models import ImageGPT           # but this works.
   ```

- ~several ways to import a certain class, for example:~ <- Shall we discuss this in #423?
   ```python
   from pl_bolts.models import LogisticRegression
   from pl_bolts.models.regression import LogisticRegression
   from pl_bolts.models.regression.logistic_regression import LogisticRegression
   ```

### Pitch
Two options I have in mind to make it consistent are:

1. Make all models available right under `pl_bolts.models`
   ```python
   pl_bolts.models.VAE
   pl_bolts.models.AE
   pl_bolts.models.ImageGPT
   pl_bolts.models.PixelCNN
   pl_bolts.models.SwAV
   pl_bolts.models.SimCLR
   ...
   ```

2. Split models into domains, e.g. `vision`, `self_supervised`... (which I prefer)
   ```python
   pl_bolts.models.autoencoders.VAE
   pl_bolts.models.autoencoders.AE
   pl_bolts.models.vision.ImageGPT
   pl_bolts.models.vision.PixelCNN
   pl_bolts.models.self_supervised.SwAV
   pl_bolts.models.self_supervised.SimCLR
   ...
   ```

@Borda @ananyahjha93 @annikabrundyn @oke-aditya I'd love to hear your opinions on if it should be as is or another interface :]

**Comment by oke-aditya** [comment]
I feel 2nd is better. I also prefer that there should be only one interface to import a model / function.

So API number 1 `from pl_bolts.models import LogisticRegression` doesn't work.

Still users can do `from pl_bolts.models.regression.logistic_regression import LogisticRegression`.  

We cannot stop this any way but this is highly discouraged. This is nature of python, people are browsing to a particular file and importing a function. This allows access to `_foobar` functions as well. Even if we don't expose `__init__` the above is still possible. (though we should ensure we haven't exposed such modules in `__init__`).
So we cannot refactor to have only 1 way of importing, there will be these 2 (inevitable).

I am not sure if we should warn people if they do above, but the possibly users should do 
`from pl_bolts.models.regression import LogisticRegression` .

**Comment by akihironitta** [comment]
@oke-aditya Thank you for sharing your thoughts!
>  we should ensure we haven't exposed such modules in `__init__`.

I agree :]
> So we cannot refactor to have only 1 way of importing, there will be these 2 (inevitable).

I know, so do you think we should rename to `_logistic_regression` which explicitly discourage users to import like `from pl_bolts.models.regression._logistic_regression import LogisticRegression`, which I suggested this renaming in #423? Indeed, Optuna, a hyperparameter optimisation library, does something like this in https://github.com/optuna/optuna/tree/master/optuna/samplers.

**Comment by oke-aditya** [comment]
Seems fine, but this is a breaking change and we should inform users about this in changelog.

**Comment by oke-aditya** [comment]
Now, I'm slightly confused now how will people access `components.py` wherever they need.

It will be possible, but components are again `_components.py` which means we are discouraging to import them (but that's not true)

I suggest to club all components and create a separate space such as

`pl_bolts.components.detection`, `pl_bolts.components.segmentation`, etc.

This will ease to test and re-use components.
E.g. we have duplicated code a lot for ResNet 3x3 blocks, Squeeze Excite blocks.
I can help in above refactor.

**Comment by akihironitta** [comment]
@oke-aditya Thanks a lot for your comments!

>Seems fine, but this is a breaking change and we should inform users about this in changelog.

Sure.

> I suggest to club all components and create a separate space such as
> 
> `pl_bolts.components.detection`, `pl_bolts.components.segmentation`, etc.

Not sure if the suggested API is a reasonable one, but it'd be great to continue to discuss refactoring components in another issue :]

**Comment by Borda** [comment]
> I suggest to club all components and create a separate space such as
> 
> `pl_bolts.components.detection`, `pl_bolts.components.segmentation`, etc.

What do the components shall include, some parts for Models?
We are eventually thinking about some restructure for v0.3

**Comment by oke-aditya** [comment]
> What do the components shall include, some parts for Models?
> We are eventually thinking about some restructure for v0.3

Hmm, yes small re-usable blocks which repeat in our codebase too often
E.g. MLP, ResNetEnocder Block,

**Comment by oke-aditya** [comment]
A few doubts that are 
1. Torchvision dependecy. We can avoid rewriting Resnet codes and user from torchvision. This has an illeffect of getting torchvision dependecy for most models. But an advantage is compatibility with torch and maintainability.
2. Varioous components that we have accumalated in places.
E.g. [Self supervised](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/self_supervised/resnets.py)
[AutoEncoders](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/autoencoders/components.py)
[GANs](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/gans/basic/components.py)
3. Probably we can group these into 2 parts. One for components and one for layers. Layers contain smaller aspects of componenets which are highly re-usable and not tied to any domain. E.g. say MLP, some varaints of Attentions.
4. Components can be encoder blocks, decoder blocks, say FPNs, Backbones, Heads.

Let's discuss further, I would be happy to contribute into this.

**Comment by akihironitta** [comment]
@oke-aditya 

> 1. Torchvision dependecy. We can avoid rewriting Resnet codes and user from torchvision. This has an illeffect of getting torchvision dependecy for most models. But an advantage is compatibility with torch and maintainability.
> 
> 2. Various components that we have accumalated in places.
>        E.g. [Self supervised](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/self_supervised/resnets.py) [AutoEncoders](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/autoencoders/components.py) [GANs](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/gans/basic/components.py)

I think we can replace components redundantly defined in Bolts with ones in torchvision since we already have a lot of components that depend on torchvision like callbacks, models, datamodules, transforms, ...

> 3. Probably we can group these into 2 parts. One for components and one for layers. Layers contain smaller aspects of componenets which are highly re-usable and not tied to any domain. E.g. say MLP, some varaints of Attentions.
> 4. Components can be encoder blocks, decoder blocks, say FPNs, Backbones, Heads.

Your suggestion sounds reasonable to me!

cc: @PyTorchLightning/bolts-contributors

**Comment by akihironitta** [comment]
As originally suggested in the description of this issue, I'd like to start working on this issue after #519 is resolved. 

> 2. Split models into domains, e.g. `vision`, `self_supervised`... (which I prefer)
> ```python
> pl_bolts.models.autoencoders.VAE
> pl_bolts.models.autoencoders.AE
> pl_bolts.models.vision.ImageGPT
> pl_bolts.models.vision.PixelCNN
> pl_bolts.models.self_supervised.SwAV
> pl_bolts.models.self_supervised.SimCLR
> ...
> ```

Specifically, the changes will be:
- to fix inconsistency, for example:
  ```python
  from pl_bolts.models.vision import SemSegment  # This works fine.   -> This will be kept as is.
  from pl_bolts.models.vision import ImageGPT    # This doesn't work, -> This will work after the fix.
  from pl_bolts.models import ImageGPT           # but this works.    -> This will not work after the fix.
  ```
- to eliminate duplicate, for example:
  ```python
  from pl_bolts.models import LogisticRegression                                  # This will not work after the fix.
  from pl_bolts.models.regression import LogisticRegression                       # This will be kept as is.
  from pl_bolts.models.regression.logistic_regression import LogisticRegression   # This will be kept as is. (We cannot do anything about this. See the comment by @oke-aditya)
  ```

Please let me know if there's more to discuss before making these changes.

**Comment by Borda** [comment]
I am for some refactoring in this direction, just I would recommend stabilizing the state e have right now as 0.3 and do this refactor in 0.4 which can be even next week :]

**Comment by akihironitta** [comment]
In addition to the refactor in my comment above https://github.com/PyTorchLightning/pytorch-lightning-bolts/issues/427#issuecomment-762032327, we should probably place `cil_main()` in `examples/` rather than in the codebase because it's an example of executable scripts rather than an API. By decoupling `cli_main()` from `pl_bolts/models`, we can make `pl_bolts/models` somewhat lighter.

**Comment by akihironitta** [comment]
Also, shall we move `LogisticRegression` out of `pl_bolts.models.regression`? The model isn't a regressor although the name includes "regression"... :laughing:

**Comment by oke-aditya** [comment]
Haha so true, it is not regression a task. But maybe put it under `pl_bolts.models.ml` ? Where we could have simple ml models.
